File: FreeLine-Android/app/src/main/java/com/freeline/app/monetization/MonetizationModels.kt
package com.freeline.app.monetization

import com.freeline.app.calls.CallAllowance
import com.freeline.app.messaging.MessageAllowance

enum class RewardType(
    val wireName: String,
    val buttonTitle: String,
    val rewardDescription: String,
) {
    TextEvents(
        wireName = "text_events",
        buttonTitle = "Watch Ad for 10 bonus texts",
        rewardDescription = "10 bonus texts",
    ),
    CallMinutes(
        wireName = "call_minutes",
        buttonTitle = "Watch Ad for 5 bonus minutes",
        rewardDescription = "5 bonus call minutes",
    ),
    ;

    companion object {
        fun fromWireName(value: String): RewardType =
            entries.firstOrNull { it.wireName == value } ?: TextEvents
    }
}

data class RewardClaimSummary(
    val callMinutesGranted: Int,
    val maxClaims: Int,
    val remainingClaims: Int,
    val textEventsGranted: Int,
    val totalClaims: Int,
)

data class RewardClaimPayload(
    val calls: CallAllowance,
    val claimedReward: RewardClaimSummary,
    val messages: MessageAllowance,
    val rewardType: RewardType,
    val tier: String,
    val trustScore: Int,
)

data class SubscriptionCatalogProduct(
    val description: String,
    val displayName: String,
    val entitlements: List<String>,
    val id: String,
    val monthlyCallCapMinutes: Int,
    val monthlySmsCap: Int,
    val priceLabel: String,
)

data class SubscriptionRecord(
    val createdAt: String,
    val entitlementKey: String,
    val expiresAt: String?,
    val id: String,
    val provider: String,
    val sourceProductId: String,
    val status: String,
    val transactionId: String,
    val updatedAt: String,
    val userId: String,
    val verifiedAt: String,
)

data class SubscriptionEntitlementState(
    val activeProducts: List<SubscriptionRecord>,
    val adFree: Boolean,
    val adsEnabled: Boolean,
    val displayTier: String,
    val numberLock: Boolean,
    val premiumCaps: Boolean,
)

data class MonetizationAllowanceBundle(
    val calls: CallAllowance,
    val messages: MessageAllowance,
)

data class SubscriptionUsagePlan(
    val dailyCallCapMinutes: Int,
    val dailySmsCap: Int,
    val description: String,
    val monthlyCallCapMinutes: Int,
    val monthlySmsCap: Int,
    val uniqueContactsDailyCap: Int,
)

data class SubscriptionStatusPayload(
    val allowances: MonetizationAllowanceBundle,
    val catalog: List<SubscriptionCatalogProduct>,
    val products: List<SubscriptionRecord>,
    val rewardClaims: RewardClaimSummary,
    val status: SubscriptionEntitlementState,
    val usagePlan: SubscriptionUsagePlan,
)

data class SubscriptionVerificationPayload(
    val allowances: MonetizationAllowanceBundle,
    val product: SubscriptionCatalogProduct,
    val status: SubscriptionEntitlementState,
    val verifiedEntitlements: List<SubscriptionRecord>,
)

data class UsageSummary(
    val callProgress: Float,
    val callsLabel: String,
    val messageProgress: Float,
    val messagesLabel: String,
    val shouldWarn: Boolean,
)

data class UsagePromptState(
    val message: String,
    val rewardType: RewardType,
)

data class RewardedAdRequest(
    val placement: String,
    val rewardType: RewardType,
)

data class InterstitialAdRequest(
    val placement: String,
)

class MonetizationApiException(
    message: String,
    val upgradePrompt: String? = null,
) : IllegalStateException(message)

File: FreeLine-iOS/Sources/Monetization/MonetizationModels.swift
import Foundation

enum AppTab: Hashable {
    case messages
    case calls
    case voicemail
    case settings
}

enum RewardType: String, Codable, CaseIterable, Identifiable {
    case callMinutes = "call_minutes"
    case textEvents = "text_events"

    var id: String { rawValue }

    var buttonTitle: String {
        switch self {
        case .textEvents:
            return "Watch Ad for 10 bonus texts"
        case .callMinutes:
            return "Watch Ad for 5 bonus minutes"
        }
    }

    var rewardDescription: String {
        switch self {
        case .textEvents:
            return "10 bonus texts"
        case .callMinutes:
            return "5 bonus call minutes"
        }
    }
}

struct RewardClaimSummary: Codable, Equatable {
    let callMinutesGranted: Int
    let maxClaims: Int
    let remainingClaims: Int
    let textEventsGranted: Int
    let totalClaims: Int
}

struct RewardClaimPayload: Codable, Equatable {
    let calls: CallAllowance
    let claimedReward: RewardClaimSummary
    let messages: MessageAllowance
    let rewardType: RewardType
    let tier: String
    let trustScore: Int
}

struct SubscriptionCatalogProduct: Codable, Equatable, Identifiable {
    let description: String
    let displayName: String
    let entitlements: [String]
    let id: String
    let monthlyCallCapMinutes: Int
    let monthlySmsCap: Int
    let priceLabel: String
}

struct SubscriptionRecord: Codable, Equatable, Identifiable {
    let createdAt: String
    let entitlementKey: String
    let expiresAt: String?
    let id: String
    let provider: String
    let sourceProductId: String
    let status: String
    let transactionId: String
    let updatedAt: String
    let userId: String
    let verifiedAt: String
}

struct SubscriptionEntitlementState: Codable, Equatable {
    let adFree: Bool
    let activeProducts: [SubscriptionRecord]
    let adsEnabled: Bool
    let displayTier: String
    let numberLock: Bool
    let premiumCaps: Bool
}

struct MonetizationAllowanceBundle: Codable, Equatable {
    let calls: CallAllowance
    let messages: MessageAllowance
}

struct SubscriptionUsagePlan: Codable, Equatable {
    let dailyCallCapMinutes: Int
    let dailySmsCap: Int
    let description: String
    let monthlyCallCapMinutes: Int
    let monthlySmsCap: Int
    let uniqueContactsDailyCap: Int
}

struct SubscriptionStatusPayload: Codable, Equatable {
    let allowances: MonetizationAllowanceBundle
    let catalog: [SubscriptionCatalogProduct]
    let products: [SubscriptionRecord]
    let rewardClaims: RewardClaimSummary
    let status: SubscriptionEntitlementState
    let usagePlan: SubscriptionUsagePlan
}

struct SubscriptionVerificationPayload: Codable, Equatable {
    let allowances: MonetizationAllowanceBundle
    let product: SubscriptionCatalogProduct
    let status: SubscriptionEntitlementState
    let verifiedEntitlements: [SubscriptionRecord]
}

struct UsageSummary {
    let callProgress: Double
    let callsLabel: String
    let messageProgress: Double
    let messagesLabel: String
    let shouldWarn: Bool
}

struct UsagePromptState: Identifiable, Equatable {
    let id = UUID()
    let message: String
    let rewardType: RewardType
}

struct RewardedAdRequest: Identifiable, Equatable {
    let id = UUID()
    let placement: String
    let rewardType: RewardType
}

struct InterstitialAdRequest: Identifiable, Equatable {
    let id = UUID()
    let placement: String
}
